software/contrib/envelope_generator.py

```diff
@@ -0,0 +1,222 @@
+from time import sleep
+from europi import oled, din, cv1, cv2, cv3, k1, k2, b1, b2, ain, OLED_WIDTH, OLED_HEIGHT
+from europi_script import EuroPiScript
+
+from time import sleep_ms, ticks_ms
+from math import log
+
+
+class EnvelopeGenerator(EuroPiScript):
+    def __init__(self):
+        super().__init__()
+        state = self.load_state_json()
+        
+        self.sustain_mode = state.get("sustain_mode", 1)
+        self.looping_mode = state.get("looping_mode", 0)
+
+        self.max_output_voltage = 10
+        
+        #Distance of envelope voltage from max voltage/0 before 'jumping' to it - prevents large logarithmic calculations
+        self.voltage_threshold = 0.1
+        
+        #Length of the longest possible envelope (not in any meaningful unit)
+        self.max_increment_factor = 256
+        self.update_increment_factor()
+        
+        #Time in ms between incrementing value of envelope
+        self.increment_delay = 1
+        
+        #0 will start a new envelope at the current value, 1 will start it from zero
+        self.retrig_mode = 0
+        
+        #Display refresh rate in ms
+        self.display_refresh_rate = 30
+        self.last_refreshed_display = self.display_refresh_rate
+
+        self.envelope_display_bounds = [0, 0, int(OLED_WIDTH), int(OLED_HEIGHT / 2)]
+        
+        self.direction = 3
```

Comment:
Add a comment to help readers by declaring what each state means.
```suggestion
        self.direction = 3  # Rising = 1, Falling = 0, Sustain = 3
```
[optional] The numbers seem to be in a strange order and it's missing "2". Consider revisiting the values used here to make it more logical like: 0 = Rising, 1 = Sustain, 2 = Falling.